### ISSUE TITLE
about training gnn model

### ISSUE THREAD

**Issue by seasonzhao** [sunericd/spatial_aging_clocks#1]
Dear Author,
May I ask whether the training process of the merfish_gnn model requires a significant amount of time? Additionally, in the scripts you provided, I only found 'train_full_gnn_model.py' but not the 'train_gnn_model.py' you mentioned. Are these two files different, or are they actually the same file? I look forward to your prompt reply.

Thank you so much.

**Comment by sunericd** [comment]
Thanks for the inquiry. The merfish_gnn models require several hours to train. The `train_full_gnn_model.py` file will train and save a final model on the entire training dataset, and this is the model that is used for all of the experiments/analyses. You can also perform cross-validation to assess model performance on the training dataset, which would be the only difference in the `train_gnn_model.py`.

**Comment by seasonzhao** [comment]
Thank you for your response. I noticed that your script does not explicitly select a GPU for training. I tried running the model on the CPU, but after an entire day, no epoch progress was printed. What should I do? Thank you.

**Comment by sunericd** [comment]
What parameters are you running the script with? The following parameters should run in ~1 day:

`python train_gnn_model.py 2 2 "T cell,NSC" "celltype" 0.0001 "balanced_mse"
`
If there is no epoch progress, the script is probably running the data/graph generation step (i.e. extracting graphs from the anndata object) which usually takes a few hours at most. You can check to see if this data is being generated at the right paths and if it takes a long time, there could be a file writing bottleneck.

**Comment by seasonzhao** [comment]
--- Parameters set directly ---
k_hop: 2
augment_hop: 2
center_celltypes: all
node_feature: celltype
learning_rate: 0.0001
loss: balanced_mse

Dataset already exists!
Finished processing test dataset
Dataset already exists!
Finished processing train dataset
482026

Yes, essentially all parameters are identical, except I selected 'all' for center cell types. I've verified the subgraph generation folders, and both train/test .pt files were created correctly. However, during the following training loop:

```
for epoch in range(1, 50):

    train(model, all_loader, criterion, optimizer,device)
    train_mse = test(model, all_loader, loss, criterion,device)
    
    if train_mse < best_mse: # if model improved then save
        best_model = copy.deepcopy(model)
        best_mse = train_mse
    
    if loss == "mse":
        print(f'Epoch: {epoch:03d}, Train MSE: {train_mse:.4f}')
    elif loss == "l1":
        print(f'Epoch: {epoch:03d}, Train L1: {train_mse:.4f}')
    elif loss == "balanced_mse":
        print(f'Epoch: {epoch:03d}, Train BMC: {train_mse:.4f}')
    elif loss == "npcc":
        print(f'Epoch: {epoch:03d}, Train NPCC: {train_mse:.4f}')
        
    training_results["epoch"].append(epoch)
    training_results["train"].append(train_mse) 
```
It takes a long time

**Comment by sunericd** [comment]
Thanks for sharing the additional details. Changing `center_celltypes` to `all` is creating 480K training graphs (instead of ~70K if using the `T cell,NSC` setting) and this is likely driving the longer training times. One thing to note is that you will get graphs centered around non-T cell/NSCs with the original settings due to the `augment_hop` parameter which will extract additional graphs centered on cells in graphs centered on either T cells / NSCs (the distribution of center cell types ends up being approximately uniform across the different cell types from this approach).

Alternatively you can set `augment_hop` to 0 or 1, which will decrease the number of training/testing graphs. Note that with `center_celltypes = 'all'` and `augment_hop` greater than zero, there will be larger representation of common cell types (e.g. neurons) in the training/testing sets.

**Comment by seasonzhao** [comment]
Thank you very much for your suggestions. I will test them and provide feedback. I just have one last question regarding your script '5F_GNN_perturbation_modeling.ipynb': I noticed you loaded two model.pth files - model.pth and all_model.pth. Are these models different in that the former was trained on T/NSC cells specifically, while the latter was trained using all cell types? I truly appreciate you taking time to respond despite your busy schedule.

**Comment by sunericd** [comment]
Sounds good, let me know if any of the issues persist!

The `model.pth` object was the model trained on just the training dataset and the `all_model.pth` was the model trained on the training+test datasets combined.

**Comment by seasonzhao** [comment]
I get it! Thank you so much!
